Haskell-Generate/GenRustJets.hs

```diff
@@ -0,0 +1,346 @@
+               | CTyProd CompactTy CompactTy
+     deriving (Eq, Ord)
+
+snakeCase :: String -> String
+snakeCase str = intercalate "_" . groupSingles $ (split . keepDelimsL . dropInitBlank . whenElt) isUpper =<< splitDigit
+ where
+  splitDigit = (split . condense . whenElt) isDigit $ str
+  groupSingles = map concat . groupBy singles
+   where
+    singles x y = null (tail x) && null (tail y)
+
+upperSnakeCase :: String -> String
+upperSnakeCase = map toUpper . snakeCase
+
+lowerSnakeCase :: String -> String
+lowerSnakeCase = map toLower . snakeCase
+
+compactTy :: Ty -> CompactTy
+compactTy = memoCataTy go
+ where
+  go One = CTyOne
+  go (Sum CTyOne CTyOne) = CTyWord 1
+  go (Sum CTyOne y) = CTyMaybe y
+  go (Sum x y) = CTySum x y
+  go (Prod (CTyWord wx) (CTyWord wy)) | wx == wy = CTyWord (wx + wy)
+  go (Prod x y) = CTyProd x y
+
+compactRustName :: CompactTy -> ShowS
+compactRustName s = showString "TypeName(b\"" . go s . showString "\")"
+ where
+  go CTyOne = showString "1"
+  go (CTyWord 1) = showString "2"
+  go (CTyWord 32) = showString "i"
+  go (CTyWord 64) = showString "l"
+  go (CTyWord 256) = showString "h"
+  go (CTyWord n) | even n = let rec = go (CTyWord (n `div` 2)) in showString "*" . rec . rec
+  go (CTyMaybe x) = showString "+1" . go x
+  go (CTySum x y) = showString "+" . go x . go y
+  go (CTyProd x y) = showString "*" . go x . go y
+
+showRustHash :: Hash256 -> Doc a
+showRustHash h = fillSep $ ((<> comma) . format <$> chunksOf 2 str_h)
+ where
+  format x = pretty $ "0x" ++ x
+  str_h = padding ++ text
+   where
+    padding = replicate (64 - length text) '0'
+    text = showHex (integerHash256 h) ""
+
+rustJetNode :: (TyC x, TyC y) => String -> JetData x y -> Doc a
+rustJetNode modname jet = vsep $
+  [ nest 4 (vsep ("pub const" <+> pretty (upperSnakeCase name) <> pretty (": JetNode<" ++ modname ++ "> = JetNode {") :
+      map (<>comma)
+      [ "name:" <+> pretty (modname ++ "JetName::" ++ name)
+      , "cmr:" <+> nest 4 (vsep
+        [ "Cmr(Midstate(["
+        , showRustHash (identityRoot (jetIdentity jet))
+        ]) <-> "]))"
+      , "source_ty:" <+> pretty (compactRustName (compactTy (unreflect tyx)) "")
+      , "target_ty:" <+> pretty (compactRustName (compactTy (unreflect tyy)) "")
+      ]))
+  , "};"
+  ]
+ where
+  name = jetName jet
+  (tyx, tyy) = reifyArrow jet
+
+rustJetCmr :: Module -> Doc a
+rustJetCmr mod = vsep $
+  [ nest 4 (vsep ("fn cmr(&self) -> Cmr {" :
+    [ nest 4 (vsep ("let bytes = match self {" :
+        map (<>comma)
+        [ nest 4 (vsep
+          [ pretty modname <> "::" <> pretty (jetName jet) <+> "=> ["
+          , showRustHash (identityRoot (jetIdentity jet))
+          ]) <-> "]"
+        | (SomeArrow jet) <- moduleJets mod
+        ]))
+    , "};"
+    , mempty
+    , "Cmr(Midstate(bytes))"
+    ]))
+  , "}"
+  ]
+ where
+  modname = rustModuleName mod
+
+rustJetTy fname getTy mod = vsep $
+  [ nest 4 (vsep (pretty ("fn "++fname++"(&self) -> TypeName {") :
```

Comment:
These should not  have `TypeName` around it

```diff
- Core::Add32 => TypeName(b"*2i"),
+ Core::Add32 => b"*2i"
```

Reply:
done.